### DISCUSSION TITLE
PGroongaのクラッシュセーフ機能はUnloggedテーブルでは無効になっているのか？

### DISCUSSION THREAD

**Discussion by komainu8** [groonga/groonga#1555] [グルカイ！]
UnloggedテーブルはPostgreSQLのWALを生成しないテーブルなので、クラッシュ時にリカバリーできない。
その代わりデータ更新時にWALを生成しないので、その分軽量に操作が可能なテーブルとなっている。

PostgreSQLのWAL生成時はWALバッファーにバッファリングしており、なるべくI/O効率が良いように作られているが、WALバッファーをストレージにフラッシュする必要があるので、その分どうしても遅くなる。
GroongaのWALもストレージに書き込みを行う関係で、クラッシュセーフがOFFのときよりONのときのほうが更新性能が落ちる。

UnloggedテーブルにもPGroongaのインデックスを設定できるが、UnloggedテーブルはPostgreSQLレベルでクラッシュセーフではないので、このテーブルまでPGroongaのクラッシュセーフ機能の対象にする必要はない。
（データが壊れたら作り直す前提のテーブルであるため、PGroongaのインデックスを自動で復旧する意味がない。）

もし、UnloggedテーブルがPGroongaのクラッシュセーフ機能の対象なら「データ更新時にWALを生成しないので、その分軽量に操作が可能なテーブル」というメリットを打ち消してしまうため、Unloggedテーブルはクラッシュセーフ機能の対象から外したほうがいい。

外す場合は、GroongaのWALを生成する処理をスキップする必要があるが、どこでどうやってGroongaのWALを生成しているのかわからないので教えてほしい。

**Comment by komainu8** [comment]
### グルカイ！第31回： https://www.youtube.com/live/lOxvr1XIMpo

UnloggedテーブルであってもPGroongaのクラッシュセーフは動いています。特にテーブルの種類によって動きを変えるようなことをしてていないためです。

UnloggedテーブルのときだけPGroongaのクラッシュセーフ機能をOFFにするにはどう実装するかですが、UnloggedテーブルかどうかはPostgreSQLのAPIを使うことでわかるので、それを使って処理を分けるような流れになります。

PGroongaはスレッドごとにロールというパラメータを設定しています。
このロールによって、WALの生成有無を決定しています。
ロールは3つあります。

* `NONE`：デフォルト。WALを生成しません。
* `SECONDARY`：更新時にWALを生成します。それ以外のことはします。クラッシュセーフがONのときのデフォルトです。
* `PRIMARY`：特別な時に使うロールです。今回の話では関係ありません。

参考資料：https://www.clear-code.com/blog/2022/3/31/pgroonga-crash-safe.html

UnloggedテーブルでWALの生成をOFFにするには、ロールを`NONE`にすればよいです。
それ以外のテーブルを操作するときはロールを`SECONDARY`にすればよいです。

PGroongaレベルでは、データ更新時にPostgreSQLからコールバックが呼ばれます。インデックス単位で呼び出されるので、そのインデックスが属するテーブルがUnloggedかどうかを判定して、Unloggedだったらロールを変更し、処理完了時にロールを元に戻します。

更新時に呼ばれるすべてのコールバックの最初と最後でロールの一時変更、復帰を実装すれよいです。
更新操作といっても、PostgreSQLは追記型なので、追記と削除しかりません。削除は`VACUUM`時です。
`SELECT`時にもデータを削除することがあるので、そのときにも実装が必要です。

無効なレコードは`VACUUM`時に削除しますが、`VACUUM`はそれなりに重い処理なので、こまめに不要なレコードを削除したいです。
検索結果をPGroongaからPostgreSQLに返す時にPostgreSQLからすでに不要なレコードを教えてもらえるので、そのタイミングで不要なレコードを削除しています。

不要なレコードかどうかはPostgreSQLしかしらないため、PostgreSQLを経由せずにPGroongaで検索した場合にすでに削除したレコードを返してしまう恐れがあります。
この問題を防ぐためにPGroongaでは、 `pgroonga_tuple_is_alive()`という関数を用意しています。
この関数を使うとPGroongaがPostgreSQLに不要なレコードかどうかを尋ねて、正しい結果を返せるようになります。
ただ、PostgreSQLに不要かどうかを尋ねるのでパフォーマンスは少し落ちる可能性があります。

参考資料：https://pgroonga.github.io/ja/reference/groonga-functions/pgroonga-tuple-is-alive.html

**Comment by askdkc** [comment]
後から来た人がGroongaを学びやすいようにグルカイ！のリンクもあった方が良いかも？

https://www.youtube.com/live/lOxvr1XIMpo

**Reply by komainu8** [reply]
追加しました！ありがとうございます！
